proposals/compile-time-dependency-resolution.md

```diff
@@ -0,0 +1,590 @@
+
+  override fun loadAll(): Map<Int, User> {
+    return storedUsers
+  }
+
+  override fun loadById(id: Int): User? {
+    return storedUsers[id]
+  }
+
+  override fun User.save() {
+    storedUsers[this.id] = this
+  }
+}
+```
+
+In KEEP, as it’s coded now, **extensions are named** primarily with the purpose of supporting Java.  We’d be fine with this narrower approach that we’re illustrating here, but would be open to iterating towards allowing definition through properties and anonymous classes if there’s a need for it.
+
+
+Now that we've got the constraint definition (interface) and a way to provide evidence of an extension for it, we just need to connect the dots. Interfaces can be used to define constraints of a function or a class constructor. We use the `with` keyword for that.
+
+```kotlin
+fun <A> fetchById(id: Int, with repository: Repository<A>): A? {
+  return loadById(id) // Repository syntax is automatically activated inside the function scope!
+}
+```
+
+As you can see, the constraint syntax is automatically active inside the function scope, so we can call its functions at will.  That's because we consider `Repository` a constraint of our program at this point. In other words, the program cannot work without it, it's a requirement. That means the following two functions would be equivalent:
+
+
+```kotlin
+// Kotlin + KEEP-87
+fun <A> fetchById(id: Int, with repository: Repository<A>): A? {
+  return loadById(id)
+}
+
+// Regular Kotlin
+fun <A> fetchById(id: Int, repository: Repository<A>): A? =
+  with (repository) {
+    return loadById(id)
+  }
+```
+
+On the call site, we could use it as follows:
+
+```kotlin
+fetchById<User>(11829) // compiles since we got evidence of a `Repository<User>` in scope.
+fetchById<Coin>(12398) // does not compile: No `Repository<Coin>` evidence defined in scope!
+```
+
+All values can be passed to functions with extension parameters, or we can omit extension parameters and let the compiler resolve the suitable extensions for them. This makes the approach really similar to the way default arguments work in Kotlin.
+
+```kotlin
+fetchById<User>(11829) // compiles since we got evidence of a `Repository<User>` in scope.
+fetchById<User>(11829, UserRepository()) // you can provide it manually.
+```
+
+When `with` is used in class constructors, it is important to **add val to extension class fields** to make sure they are accessible in the scope of the class. Here, the `with` keyword adds the value to the scope of every method in the class. To showcase this, let's say we have a `Validator<A>`, like:
+
+```kotlin
+interface Validator<A> {
+  fun A.isValid(): Boolean
+}
+```
+
+In this scenario, the following classes would be equivalent:
+
+```kotlin
+data class Group<A>(val values: List<A>)
+
+// Kotlin + KEEP-87
+extension class ValidatedRepository<A>(with val V: Validator<A>) : Repository<A> {
+
+    val storedUsers: MutableMap<Int, A> = mutableMapOf() // e.g. users stored in a DB
```

Comment:
`ValidatedRepository<A>  : Repository<A>`, we shouldn't talk about users -> `storedUsers`